### PR TITLE
feat: Add a `mut` annotation to struct fields

### PR DESCRIPTION
**Pull request by dghosef**
#8308

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 1]
```diff
@@ -1207,6 +1207,23 @@ object ResolutionError {
     }
   }
 
+  /**
+   * An error raised to indicate a `put` struct expression attempts to modify an immutable field
+   *
+   * @param loc    the location where the error occurred
+   */
+  case class IllegalPut(loc: SourceLocation) extends ResolutionError with Recoverable {
+    override def summary: String = s"Modification of struct field not marked as `mut`"
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:1216]: How about:

Modification of immutable field. Mark the field as `mut` to allow mutation.

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 2]
```diff
@@ -1207,6 +1207,23 @@ object ResolutionError {
     }
   }
 
+  /**
+   * An error raised to indicate a `put` struct expression attempts to modify an immutable field
+   *
+   * @param loc    the location where the error occurred
+   */
+  case class IllegalPut(loc: SourceLocation) extends ResolutionError with Recoverable {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:1215]: Also, lets get the field as part of the error.

Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:1215]: IllegalPut -> NonModifiableField  or something like it

**Comment by dghosef** [comment]
I'll look at those suggestions.

However, I've started working on lookup of struct(which is related to `pub`) and that one would conflict with this, so I will get that one in and then finish this one

**Comment by magnus-madsen** [comment]
Let me know when I should look at this again.

**Comment by dghosef** [comment]
You can look at it again. I feel this and #8334 should remain drafts until we discuss on Monday

**Pull request closed without merging**